// sanity-plugin-media/lib/styled/Box.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var styled_system_1 = require("styled-system");
var styled_components_1 = __importDefault(require("styled-components"));
var textColor = styled_system_1.style({
    prop: 'textColor',
    cssProperty: 'color',
    key: 'colors'
});
var Box = styled_components_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  box-sizing: border-box;\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n\n  ", "\n"], ["\n  box-sizing: border-box;\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n  ", ";\n\n  ",
    "\n"])), styled_system_1.border, styled_system_1.color, styled_system_1.flexbox, styled_system_1.grid, styled_system_1.layout, styled_system_1.position, styled_system_1.space, textColor, styled_system_1.typography, styled_system_1.system({
    boxShadow: true,
    boxSizing: true,
    cursor: true,
    gridColumnGap: {
        property: 'gridColumnGap',
        scale: 'space'
    },
    gridGap: {
        property: 'gridGap',
        scale: 'space'
    },
    gridRowGap: {
        property: 'gridRowGap',
        scale: 'space'
    },
    gridRowEnd: true,
    gridRowStart: true,
    gridTemplateColumns: {
        property: 'gridTemplateColumns',
        scale: 'gridTemplateColumns'
    },
    outline: true,
    overflowY: true,
    pointerEvents: true,
    textOverflow: true,
    textTransform: true,
    transform: true,
    transition: true,
    userSelect: true,
    whiteSpace: true
}));
exports.default = Box;
var templateObject_1;
