// sanity-plugin-media/lib/styled/FormInput.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var styled_components_1 = __importDefault(require("styled-components"));
var Box_1 = __importDefault(require("./Box"));
var FormInput = styled_components_1.default(Box_1.default)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  appearance: none;\n  background: transparent;\n  border-style: solid;\n  border-width: 1px;\n  outline: none;\n  transition: 250ms all;\n\n  @media (hover: hover) and (pointer: fine) {\n    &:hover:enabled:not(:focus) {\n      background-color: rgba(255, 255, 255, 0.02);\n    }\n  }\n\n  &:focus {\n    border-color: ", ";\n  }\n\n  &:disabled {\n  }\n\n  ::placeholder {\n    color: ", ";\n  }\n\n  ::selection {\n    background: ", ";\n    color: ", ";\n  }\n"], ["\n  appearance: none;\n  background: transparent;\n  border-style: solid;\n  border-width: 1px;\n  outline: none;\n  transition: 250ms all;\n\n  @media (hover: hover) and (pointer: fine) {\n    &:hover:enabled:not(:focus) {\n      background-color: rgba(255, 255, 255, 0.02);\n    }\n  }\n\n  &:focus {\n    border-color: ", ";\n  }\n\n  &:disabled {\n  }\n\n  ::placeholder {\n    color: ", ";\n  }\n\n  ::selection {\n    background: ", ";\n    color: ", ";\n  }\n"])), function (props) { return props.theme.colors.gray; }, function (props) { return props.theme.colors.gray; }, function (props) { return props.theme.colors.black; }, function (props) { return props.theme.colors.white; });
FormInput.defaultProps = {
    borderColor: 'darkGray',
    borderRadius: '2px',
    fontSize: 1,
    lineHeight: 1.5,
    py: '6px',
    pl: 2,
    pr: 2,
    textColor: 'lightGray'
};
exports.default = FormInput;
var templateObject_1;
