// sanity-plugin-media/lib/utils/validateConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var joi_1 = __importDefault(require("@hapi/joi"));
var customField = joi_1.default.object({
    description: joi_1.default.string(),
    name: joi_1.default.string().required(),
    title: joi_1.default.string().required(),
    type: joi_1.default.string()
        .regex(/^string$/)
        .required()
});
var filter = joi_1.default.object({
    title: joi_1.default.string().required(),
    value: joi_1.default.string().required()
});
var schema = joi_1.default.object({
    customFields: joi_1.default.array().items(customField),
    filters: joi_1.default.array().items(filter)
});
function validateConfig(config) {
    var options = {
        abortEarly: false,
        stripUnknown: true
    };
    return schema.validate(config, options);
}
exports.default = validateConfig;
