// sanity-plugin-media/lib/modules/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rootReducer = exports.rootEpic = void 0;
var redux_1 = require("redux");
var redux_observable_1 = require("redux-observable");
var assets_1 = __importStar(require("./assets"));
var dialog_1 = __importStar(require("./dialog"));
var document_1 = __importDefault(require("./document"));
var snackbars_1 = __importStar(require("./snackbars"));
exports.rootEpic = redux_observable_1.combineEpics(assets_1.assetsDeleteEpic, assets_1.assetsDeletePickedEpic, assets_1.assetsFetchEpic, assets_1.assetsFetchNextPageEpic, assets_1.assetsFetchPageEpic, assets_1.assetsFetchPageIndexEpic, dialog_1.dialogClearEpic, dialog_1.dialogShowConflictsEpic, snackbars_1.snackbarsAddDeleteErrorsEpic, snackbars_1.snackbarsAddFetchErrorEpic, snackbars_1.snackbarsAddSuccessEpic);
exports.rootReducer = redux_1.combineReducers({
    assets: assets_1.default,
    dialog: dialog_1.default,
    document: document_1.default,
    snackbars: snackbars_1.default
});
